### PR TITLE
Restructure modules

### LINKED ISSUE #31: dev: Restructure modules
It'd be more logical to partition `kernels.py` out into one module per kernel.

### PR DESCRIPTION
**Pull request by thomaspinder**
This PR restructures JaxKern so that kernels are broken out by their type. Further, computation objects are given their own modules.

Resolves #31

### CODE REVIEW

**Comment by Daniel-Dodd** [comment]
Do we need this `py.typed` file?

**Comment by Daniel-Dodd** [comment]
Suggestion: While we (or you) are refactoring, would it be sensible to structure the kernels and the computations across two submodules rather than in the main directory? I see the computations growing in size in the near future (inter-domain, Kronecker, Markovian dynamics, ...). It might be nice to have the computations split up into separate files for better readability (like I did in `JaxLinOp`)? You could also consider have submodules within the kernels (for e.g. stationary). It might make things easier to maintain/ reduce the amount of scrolling? Interested to hear your thoughts on this.

**Comment by thomaspinder** [comment]
> Do we need this `py.typed` file?

Not right now. Removed.

**Comment by thomaspinder** [comment]
> Suggestion: While we (or you) are refactoring, would it be sensible to structure the kernels and the computations across two submodules rather than in the main directory? I see the computations growing in size in the near future (inter-domain, Kronecker, Markovian dynamics, ...). It might be nice to have the computations split up into separate files for better readability (like I did in `JaxLinOp`)? You could also consider have submodules within the kernels (for e.g. stationary). It might make things easier to maintain/ reduce the amount of scrolling? Interested to hear your thoughts on this.

Yeah nice idea.

**Comment by thomaspinder** [comment]
> > Suggestion: While we (or you) are refactoring, would it be sensible to structure the kernels and the computations across two submodules rather than in the main directory? I see the computations growing in size in the near future (inter-domain, Kronecker, Markovian dynamics, ...). It might be nice to have the computations split up into separate files for better readability (like I did in `JaxLinOp`)? You could also consider have submodules within the kernels (for e.g. stationary). It might make things easier to maintain/ reduce the amount of scrolling? Interested to hear your thoughts on this.
> 
> Yeah nice idea.

@Daniel-Dodd  - this is now done.

**Review by Daniel-Dodd — COMMENTED** [review]
Sorry @thomaspinder, I don't think might have see my review comments. :)

**Comment by Daniel-Dodd** [comment]
Sorry @thomaspinder these comment are a little out of date with regards to their file references.

File: jaxkern/computations.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        """
+
+        diag = vmap(lambda x: self.kernel_fn(params, x, x))(inputs)
+
+        return DiagonalLinearOperator(diag=diag)
+
+    def cross_covariance(
+        self, params: Dict, x: Float[Array, "N D"], y: Float[Array, "M D"]
+    ) -> Float[Array, "N M"]:
+        raise ValueError("Cross covariance not defined for diagonal kernels.")
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/computations.py]: This is fine - there is `gram` method overrided defined that will take care of things.

File: jaxkern/computations.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    def cross_covariance(
+        self, params: Dict, x: Float[Array, "N D"], y: Float[Array, "M D"]
+    ) -> Float[Array, "N M"]:
+        raise ValueError("Cross covariance not defined for constant diagonal kernels.")
+
+
+class EigenKernelComputation(AbstractKernelComputation):
+    def __init__(
+        self,
+        kernel_fn: Callable[
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/computations.py]: Will open an issue. This should be a separate PR.

File: jaxkern/non_euclidean.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from .computations import EigenKernelComputation
+from .nonstationary import AbstractKernel
+from .utils import jax_gather_nd
+
+
+##########################################
+# Graph kernels
+##########################################
+class GraphKernel(AbstractKernel):
+    def __init__(
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/non_euclidean.py]: Will open an issue

File: jaxkern/stationary.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+from .base import AbstractKernel
+from .computations import (
+    AbstractKernelComputation,
+    DenseKernelComputation,
+)
+from .utils import euclidean_distance, squared_distance
+
+
+class RBF(AbstractKernel):
+    """The Radial Basis Function (RBF) kernel."""
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/stationary.py]: I'd err against over-subclassing. Stationarity is a property of a kernel, not a type, so I'm not even sure subclassing makes sense here. Practically speaking, it makes no difference to my RFF implementation, if anything, maybe more difficult.

File: jaxkern/computations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply jaxkern/computations.py]: Really? This is beyond the scope of this PR. Will open bug report.
```
import jaxkern as jk
import  jax.numpy as jnp

from jaxkern.stationary.white import White

w = White()
p = w.init_params(None)

w.cross_covariance(p, jnp.array([1,2,3]), jnp.array([4,5,6]))

# ValueError: Cross covariance not defined for constant diagonal kernels.
```

Also the gram matrix seems to be broken on the `White` kernel (get an error of `ValueError: `value` must be one dimensional scalar, but 'value.shape = ()'`. Yet again beyond the scope of this PR, will open bug report.

[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply jaxkern/computations.py]: Gram:

```python
import jaxkern as jk
import  jax.numpy as jnp

from jaxkern.stationary.white import White

w = White()
p = w.init_params(None)

w.gram(p, jnp.array([[1],[2],[3]]))

# ValueError: `value` must be one dimensional scalar, but `value.shape = ()`.
```

Note this is not a not a bug in JaxLinOp. The array needs reshaping or JaxLinOp needs to alter its shape convention.

**Pull request merged**